src/main/scala/Validator.scala

```diff
@@ -1,115 +1,387 @@
+          ""
+        )
+      )
+    } else {
+      allPrimaryKeyValues += primaryKeyValues
+      None
+    }
+  }
+
+  private def setParentTableForeignKeyReferences(
+      validateRowOutput: ValidateRowOutput,
+      parentTableForeignKeyReferences: Map[ParentTableForeignKeyReference, Set[
+        KeyWithContext
+      ]]
+  ) = {
+    validateRowOutput.parentTableForeignKeyReferences
+      .foreach {
+        case (k, value) => {
+          val allPossibleParentKeyValues =
+            parentTableForeignKeyReferences.getOrElse(
+              k,
+              Set[KeyWithContext]()
+            )
+          if (allPossibleParentKeyValues.contains(value)) {
+            allPossibleParentKeyValues -= value
+            value.isDuplicate = true
+          }
+          allPossibleParentKeyValues += value
+          parentTableForeignKeyReferences(k) = allPossibleParentKeyValues
+        }
+      }
+  }
+
+  private def setChildTableForeignKeys(
+      validateRowOutput: ValidateRowOutput,
+      childTableForeignKeys: Map[ChildTableForeignKey, Set[KeyWithContext]]
+  ) = {
+    validateRowOutput.childTableForeignKeys
+      .foreach {
+        case (k, value) => {
+          val childKeyValues = childTableForeignKeys.getOrElse(
+            k,
+            Set[KeyWithContext]()
           )
+          childKeyValues += value
+          childTableForeignKeys(k) = childKeyValues
         }
-        val result = table.validateRow(row)
-        result match {
-          case Some(validateRowOutput) => {
-            val primaryKeyValues = validateRowOutput.primaryKeyValues
-            if (allPrimaryKeyValues.contains(primaryKeyValues)) {
-              errors = errors :+ ErrorWithCsvContext(
-                "duplicate_key",
+      }
+  }
+
+  private def validateForeignKeyReferences(
+      childTableForeignKeysByTable: Map[
+        Table,
+        Map[ChildTableForeignKey, Set[KeyWithContext]]
+      ],
+      parentTableForeignKeyReferencesByTable: Map[
+        Table,
+        Map[ParentTableForeignKeyReference, Set[KeyWithContext]]
+      ]
+  ): Errors = {
+    // Child Table : Parent Table
+    // Country, Year, Population  : Country, Name
+    // UK, 2021, 67M  : UK, United Kingdom
+    // EU, 2021, 448M : EU, Europe
+    var errors: Errors = Array[ErrorWithCsvContext]()
+    for (
+      (parentTable, mapParentTableForeignKeyReferenceToAllPossibleValues) <-
+        parentTableForeignKeyReferencesByTable
+    ) {
+      for (
+        (parentTableForeignKeyReference, allPossibleParentTableValues) <-
+          mapParentTableForeignKeyReferenceToAllPossibleValues
+      ) {
+        val childTableForeignKeys
+            : Map[ChildTableForeignKey, Set[KeyWithContext]] =
+          childTableForeignKeysByTable
+            .get(parentTableForeignKeyReference.childTable)
+            .getOrElse(
+              throw new Exception(
+                s"Could not find corresponding child table for parent table ${parentTable.url}"
```

Comment:
I think we really want to mention what the child table is here in the exception too.

Reply:
okay